File: js/src/test/cloneModels.test.ts
import { existsSync, readFileSync, mkdtempSync, rmSync } from "node:fs";
import { join } from "node:path";
import { tmpdir } from "node:os";
import test from "node:test";
import assert from "node:assert/strict";
import { Engine } from "../engine.js";
import { Enroller, profileFrom } from "../enroll.js";
import { saveVoice } from "../voice.js";
import { refuseIfAssetsRequired } from "./assets.js";

const {LOUDKIT_CKPT: first, LOUDKIT_ONNX_DIR: graphs, LOUDKIT_SECOND_CKPT: second,
  LOUDKIT_SECOND_ONNX_DIR: secondGraphs, LOUDKIT_TOKENIZER: tokenizer} = process.env;
const available = [first,graphs,second,secondGraphs,tokenizer].every(p => p && existsSync(p));
refuseIfAssetsRequired(available, "LOUDKIT_CKPT/SECOND_CKPT/ONNX_DIR/SECOND_ONNX_DIR/TOKENIZER are not all present");

test("enroll once and reuse the unchanged profile through both models", {skip:!available && "set both model paths and tokenizer"}, async () => {
  const bytes = readFileSync(join(process.env.LOUDKIT_ENROLL_FIXTURE ?? "../tests/data/enrollment", "ref_audio.f32"));
  const view = new DataView(bytes.buffer,bytes.byteOffset,bytes.byteLength);
  const samples = Float32Array.from({length:bytes.length/4},(_,i) => view.getFloat32(i*4,true));
  const enr = await Enroller.load(graphs!, {onnxProvider:"cpu"});
  let profile;
  try { profile = profileFrom(await enr.enroll(samples,24000),{name:"shared",sampleRate:24000,language:"en"}); }
  finally { await enr.close(); }
  const dir = mkdtempSync(join(tmpdir(),"loudkit-shared-voice-"));
  try {
    const path = join(dir,"shared.voice.safetensors");
    saveVoice(profile,path);
    const before = readFileSync(path);
    const modes = new Set<string>();
    for (const [checkpoint,onnx] of [[first,graphs],[second,secondGraphs]]) {
      const engine = await Engine.loadPaths(checkpoint!,onnx!,tokenizer!,{onnxProvider:"cpu"});
      try {
        modes.add(engine.config.decode ?? "single");
        const result = await engine.synthesize("Hello, this is a shared voice.",profile,{seed:7,language:"en"});
        assert.ok(result.audio.length > 0);
        assert.ok(result.audio.every(Number.isFinite));
      } finally { await engine.close(); }
    }
    assert.deepEqual(modes,new Set(["single","fusion_mtp2"]));
    saveVoice(profile,path);
    assert.deepEqual(readFileSync(path),before);
  } finally { rmSync(dir,{recursive:true,force:true}); }
});

File: go/conformance/clone_models_test.go
package conformance

import (
	"bytes"
	"math"
	"os"
	"path/filepath"
	"testing"

	"github.com/loudreader/loudkit/go/config"
	"github.com/loudreader/loudkit/go/engine"
	"github.com/loudreader/loudkit/go/enroll"
	"github.com/loudreader/loudkit/go/onnx"
)

// One enrollment result must remain reusable across both decode families.
func TestCloneOnceAcrossBothModels(t *testing.T) {
	first, second := os.Getenv("LOUDKIT_CKPT"), os.Getenv("LOUDKIT_SECOND_CKPT")
	graphs, secondGraphs := os.Getenv("LOUDKIT_ONNX_DIR"), os.Getenv("LOUDKIT_SECOND_ONNX_DIR")
	lib := os.Getenv("LOUDKIT_ONNXRUNTIME_LIB")
	if first == "" || second == "" || graphs == "" || secondGraphs == "" || lib == "" {
		skipOrFail(t, "set both model checkpoint/graph paths and ONNX runtime library")
	}
	fixture := os.Getenv("LOUDKIT_FIXTURE_DIR")
	if fixture == "" {
		fixture = "../../tests/data/conformance"
	}
	enrollment := os.Getenv("LOUDKIT_ENROLL_FIXTURE")
	if enrollment == "" {
		enrollment = "../../tests/data/enrollment"
	}
	onnx.SetSharedLibraryPath(lib)
	if err := onnx.InitializeEnvironment(); err != nil {
		t.Fatal(err)
	}
	defer onnx.DestroyEnvironment()
	execution := config.ExecutionConfig{ONNXProvider: config.ProviderCPU}
	enr, err := enroll.LoadEnrollerWith(graphs, execution)
	if err != nil {
		t.Fatal(err)
	}
	result, err := enr.Enroll(readF32(t, filepath.Join(enrollment, "ref_audio.f32")), 24000)
	enr.Close()
	if err != nil {
		t.Fatal(err)
	}
	profile := result.Profile("shared", 24000, "en")
	path := filepath.Join(t.TempDir(), "shared.voice.safetensors")
	if err := profile.Save(path); err != nil {
		t.Fatal(err)
	}
	before, err := os.ReadFile(path)
	if err != nil {
		t.Fatal(err)
	}
	modes := map[string]bool{}
	for i, checkpoint := range []string{first, second} {
		dir := []string{graphs, secondGraphs}[i]
		eng, err := engine.LoadWith(checkpoint, dir, filepath.Join(fixture, "tokenizer.json"), execution)
		if err != nil {
			t.Fatal(err)
		}
		modes[eng.Config().DecodeMode] = true
		speech, err := eng.Synthesize("Hello, this is a shared voice.", profile, engine.Options{Seed: 7, Language: "en"})
		eng.Close()
		if err != nil {
			t.Fatal(err)
		}
		if len(speech.Audio) == 0 {
			t.Fatal("empty speech")
		}
		for _, sample := range speech.Audio {
			if math.IsNaN(float64(sample)) || math.IsInf(float64(sample), 0) {
				t.Fatal("non-finite speech")
			}
		}
	}
	if !modes["single"] || !modes["fusion_mtp2"] {
		t.Fatal("the test requires one model from each decode family")
	}
	if err := profile.Save(path); err != nil {
		t.Fatal(err)
	}
	after, err := os.ReadFile(path)
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Equal(before, after) {
		t.Fatal("synthesis mutated the shared voice profile")
	}
}
